languages/ruby/exercises/concept/enumeration/.docs/introduction.md

```diff
@@ -0,0 +1,42 @@
+Enumeration is the act of stepping through a collection (Array, Hash, etc) and performing some action on each object.
+Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.
+
+A simple enumeration to print each word in an array would look like this:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each do |word| 
+  print word 
+end
+
+#=> the
+#=> cat
+#=> ...
+#=> mat
+```
+
+In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and prints it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchangeable and differing opinions exist about when each should be used.
+
+We can also chain enumerable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each.with_index { |word, index| print "#{index}: #{word}" }
+
+#=> 0. the
+#=> 1. cat
+#=> ...
+#=> 5. mat
+```
+
+Enumerating hashes is exactly the same as enumerating arrays, except that the block receives two parameters: the key and the value:
```

Comment:
```suggestion
Enumerating `Hash` objects is exactly the same as enumerating `Array` objects, except that the block receives two parameters: the key and the value:
```

In Ruby (and computer science) the word "hash" is used.  Here we are specifically speaking of `Hash` objects, as opposed to the `hash` that is required to be callable on anything that can be a key for a `Hash`.  We don't want to mention that here, but we also don't want to confuse the terms, so being explicit that we are iterating a `Hash` is better than speaking loosely about it.  And since the documentation is linked, the documentation also speaks about "hash" (it is a method as well), and it talks about how or why the hash is used.